azurelinuxagent/common/osutil/default.py

```diff
@@ -904,26 +906,15 @@ def get_primary_interface(self):
 
         if not self.disable_route_warning:
             logger.info("Examine /proc/net/route for primary interface")
-        with open('/proc/net/route') as routing_table:
-            idx = 0
-            for header in filter(lambda h: len(h) > 0, routing_table.readline().strip(" \n").split("\t")):
-                if header == hdr_iface:
-                    idx_iface = idx
-                elif header == hdr_dest:
-                    idx_dest = idx
-                elif header == hdr_flags:
-                    idx_flags = idx
-                elif header == hdr_metric:
-                    idx_metric = idx
-                idx = idx + 1
-            for entry in routing_table.readlines():
-                route = entry.strip(" \n").split("\t")
-                if route[idx_dest] == DEFAULT_DEST and int(route[idx_flags]) & RTF_GATEWAY == RTF_GATEWAY:
-                    metric = int(route[idx_metric])
-                    iface = route[idx_iface]
-                    if primary is None or metric < primary_metric:
-                        primary = iface
-                        primary_metric = metric
+        route_table = DefaultOSUtil.read_route_table()
+        route_list = DefaultOSUtil.get_list_of_routes(route_table)
+        for route in route_list:
```

Comment:
This is a little longer, but it'll definitely be faster (and better with memory) for even moderately-sized route tables. I also think it's far clearer.

There may already be an `is_default_route()` function defined elsewhere; if so, just use that instead of making a new one. Alternatively, just use a lambda in the filter() call instead of defining the function elsewhere, if you want to save a line.

```python
route_table = DefaultOSUtil.read_route_table()
def is_default(route):
    return route[idx_dest] == DEFAULT_DEST and int(route[idx_flags]) & RTF_GATEWAY == RTF_GATEWAY
candidates = filter(DefaultOSUtil.get_list_of_routes(route_table), is_default)
if len(candidates) > 0:
    def get_metric(route):
        return int(route[idx_metric])
    primary_route = min(candidates, get_metric)
    primary = primary_route[idx_face]
    primary_metric = get_metric(primary_route)
```

Reply:
done, thanks for the advice
just one thing, the route is now a RouteEntry object and it does not support indexing, so we need to access its parameters instead; this is reflected in the code change